lib/rucio/core/authentication.py

```diff
@@ -61,14 +89,83 @@ def generate_key(token, session=None):
     expiration_time=3600
 )
 
+# private/protected file containing Rucio Client secrets known to the Identity Provider as well
+idpsecrets = config_get('oidc', 'idpsecrets')
+
+# TO-DO permission layer: if scope == 'wlcg.groups'
+# --> check 'profile' info (requested profile scope)
+# decide on exchanges permitted and save the dict per token in the DB
+# (user groups maybe need to be transformed in to exchange token patameters
+# and the permitted token echange audience and scopes to be saved in the DB table as well)
+exchange_params = {
+    'audiences': {'rucio': 'fts'},
+    'scopes': {}
+}
+
+
+def filter_thread_load(query, column_name, total_workers, worker_number, session=None):
+    """
+    Adds filter to a DB session query in order to distribute
+    the queried number of rows among all workers.
+    :param total_workers: total number of threads started
+    :param worker_number: the number of the worker asking for query results
+    :param column: name of the column w.r.t. which the partitioning should be made
+    :param query: the session query
+
+    :returns: session query in case all went OK, Exception otherwise.
+
+    """
+    try:
+        if worker_number and total_workers and total_workers - 1 > 0:
+            if session.bind.dialect.name == 'oracle':
+                bindparams = [bindparam('worker_number', worker_number - 1), bindparam('total_workers', total_workers - 1)]
+                query = query.filter(text('ORA_HASH(%s, :total_workers) = :worker_number' % column_name, bindparams=bindparams))
+            elif session.bind.dialect.name == 'mysql':
+                query = query.filter(text('mod(md5(%s), %s) = %s' % (column_name, total_workers - 1, worker_number - 1)))
+            elif session.bind.dialect.name == 'postgresql':
+                query = query.filter(text('mod(abs((\'x\'||md5(%s))::bit(32)::int), %s) = %s' % (column_name, total_workers - 1, worker_number - 1)))
+
+        return query
+
+    except Exception as error:
+        raise RucioException(error.args)
+
+
+def query_bunches(query, bunch_by):
```

Comment:
i think we have a function for this in common/utils already?

Reply:
Checking now, I did not find any. Please, let me know which one if you find it.